primitives/forest-verifier/src/lib.rs

```diff
@@ -238,15 +246,33 @@ where
         let mut trie = TrieDBMutBuilder::<T>::from_existing(&mut memdb, &mut root).build();
 
         // Apply mutations to the trie
+        let mut mutated_keys_and_values: Vec<(Self::Key, Option<Vec<u8>>)> = Vec::new();
         for mutation in mutations.iter() {
             match mutation {
-                (key, TrieMutation::Add(_)) => {
-                    trie.insert(key.as_ref(), &[])
+                (key, TrieMutation::Add(mutation)) => {
+                    trie.insert(key.as_ref(), &mutation.value)
                         .map_err(|_| "Failed to insert key into trie.")?;
+                    mutated_keys_and_values.push((key.clone(), Some(mutation.value.clone())));
                 }
                 (key, TrieMutation::Remove(_)) => {
-                    trie.remove(key.as_ref())
+                    let node_value = trie
+                        .remove(key.as_ref())
                         .map_err(|_| "Failed to remove key from trie.")?;
+                    let previous_value = if let Some(node_value) = node_value {
+                        match node_value {
+                            trie_db::Value::Inline(value) => Some(value.to_vec()),
+                            trie_db::Value::Node(value_hash) => {
+                                let value = trie
+                                    .get(value_hash.as_ref())
+                                    .map_err(|_| "Failed to get value from trie.")?;
+                                value.map(|v| v.to_vec())
+                            }
+                            trie_db::Value::NewNode(_, value) => Some(value.to_vec()),
+                        }
```

Comment:
I believe you can do this.

```suggestion
                        match node_value {
                            trie_db::Value::Inline(value) | trie_db::Value::NewNode(_, value) => Some(value.to_vec()),
                            trie_db::Value::Node(value_hash) => {
                                let value = trie
                                    .get(value_hash.as_ref())
                                    .map_err(|_| "Failed to get value from trie.")?;
                                value.map(|v| v.to_vec())
                            }
                        }
```